src/main/java/org/springframework/data/couchbase/core/convert/join/N1qlJoinResolver.java

```diff
@@ -111,10 +120,94 @@ public static String buildQuery(ReactiveCouchbaseTemplate template, String colle
 		return statementSb.toString();
 	}
 
-	public static <R> List<R> doResolve(ReactiveCouchbaseTemplate template, String collectionName,
+	static KeySpacePair getKeySpacePair(String bucketName, String scope, String collection,
+			N1qlJoinResolverParameters parameters) {
+		Class<?> lhsClass = parameters.getEntityTypeInfo().getActualType().getType();
+		String lhScope = scope != null ? scope : getScope(lhsClass);
+		String lhCollection = collection != null ? collection : getCollection(lhsClass);
+		Class<?> rhsClass = parameters.getAssociatedEntityTypeInfo().getActualType().getType();
+		String rhScope = getScope(rhsClass);
+		String rhCollection = getCollection(rhsClass);
+		if (lhCollection != null && rhCollection != null) {
+			// they both have non-default collections
+			// It's possible that the scope for the lhs was set with an annotation on a repository method,
+			// the entity class or the repository class or a query option. Since there is no means to set
+			// the scope of the associated class by the method, repository class or query option (only
+			// the annotation) we assume that the (possibly) dynamic scope of the entity would be a better
+			// choice as it is logical to put collections to be joined in the same scope. Note that lhScope
+			// is used for both keyspaces.
+			return new KeySpacePair(lhCollection,
+					x(x("default") + ":" + i(bucketName) + "." + i(lhScope) + "." + i(lhCollection)), //
```

Comment:
is the default: needed? I think n1ql can infer that if not provided? Maybe I'm wrong though..

Reply:
Your are correct. Removed.